client/sniffer/src/main/java/org/elasticsearch/client/sniff/ElasticsearchHostsSniffer.java

```diff
@@ -89,68 +92,110 @@ public ElasticsearchHostsSniffer(RestClient restClient, long sniffRequestTimeout
-                                    boundAddressAsURI.getScheme());
+                            URI publishAddressAsURI = URI.create(scheme + "://" + parser.getValueAsString());
+                            publishedHost = new HttpHost(publishAddressAsURI.getHost(), publishAddressAsURI.getPort(),
+                                    publishAddressAsURI.getScheme());
+                        } else if (parser.currentToken() == JsonToken.START_ARRAY && "bound_address".equals(parser.getCurrentName())) {
+                            while (parser.nextToken() != JsonToken.END_ARRAY) {
+                                URI boundAddressAsURI = URI.create(scheme + "://" + parser.getValueAsString());
+                                boundHosts.add(new HttpHost(boundAddressAsURI.getHost(), boundAddressAsURI.getPort(),
+                                        boundAddressAsURI.getScheme()));
+                            }
                         } else if (parser.getCurrentToken() == JsonToken.START_OBJECT) {
                             parser.skipChildren();
                         }
                     }
                 } else {
                     parser.skipChildren();
                 }
+            } else if (parser.currentToken() == JsonToken.START_ARRAY) {
+                if ("roles".equals(fieldName)) {
+                    sawRoles = true;
+                    while (parser.nextToken() != JsonToken.END_ARRAY) {
+                        switch (parser.getText()) {
+                        case "master":
+                            master = true;
+                            break;
+                        case "data":
+                            data = true;
+                            break;
+                        case "ingest":
+                            ingest = true;
+                            break;
+                        default:
+                            logger.warn("unknown role [" + parser.getText() + "] on node [" + nodeId + "]");
+                        }
+                    }
+                } else {
+                    parser.skipChildren();
+                }
+            } else if (parser.currentToken().isScalarValue()) {
+                if ("version".equals(fieldName)) {
+                    version = parser.getText();
+                }
             }
         }
         //http section is not present if http is not enabled on the node, ignore such nodes
-        if (httpHost == null) {
+        if (publishedHost == null) {
             logger.debug("skipping node [" + nodeId + "] with http disabled");
-            return null;
+        } else {
+            logger.trace("adding node [" + nodeId + "]");
+            assert sawRoles : "didn't see roles for [" + nodeId + "]";
+            hosts.add(publishedHost);
+            HostMetadata meta = new HostMetadata(version, new Roles(master, data, ingest));
+            for (HttpHost bound: boundHosts) {
+                hostMetadata.put(bound, meta);
```

Comment:
Because we don't know how the client is configured. In particular, at least on my machine, our REST tests are configured against an ipv4 address but the published host is an ipv6 address.

You don't need both if you are using the sniffer to sniff the hosts, but you need it if you are using the sniffer to sniff metadata for some hosts you already have.

Reply:
I see, can you add a comment to make sure we don't forget about the why here?